### PR TITLE
Metadata service: refresh Metatron certs every 5 minutes

### PR DESCRIPTION
**Pull request by rgulewich**
This prevents the task identity document being signed with a stale certificate.

### CODE REVIEW

**Comment by sargun** [comment]
Can you wire up SIGHUP as well?

File: cmd/titus-metadata-service/main.go [hunk 1]
```diff
@@ -78,6 +80,17 @@ func readTaskConfigFile(taskID string) (*titus.ContainerInfo, error) {
 	return &cInfo, nil
 }
 
+func reloadSigner(ms *metadataserver.MetadataServer) {
+	for {
+		time.Sleep(certRefreshTime)
```
Comment (sargun) [inline cmd/titus-metadata-service/main.go]:
```
t := time.NewTicker(5 * time.Second)
defer t.Stop()
for range <-t.C {
...
}
```

File: metadataserver/identity.go [hunk 2]
```diff
@@ -29,6 +31,39 @@ func (ms *MetadataServer) generateTaskIdentity() *titus.TaskIdentity {
...
+			"newIssuer":       newCert.Issuer,
+			"newNotBefore":    newCert.NotBefore,
+			"newNotAfter":     newCert.NotAfter,
+			"newSerialNumber": newCert.SerialNumber,
+			"oldSubject":      oldCert.Subject,
+			"oldIssuer":       oldCert.Issuer,
+			"oldNotBefore":    oldCert.NotBefore,
+			"oldNotAfter":     oldCert.NotAfter,
+			"oldSerialNumber": oldCert.SerialNumber,
+		}).Infof("signer cert updated")
```
Comment (sargun) [inline metadataserver/identity.go]: why infof?

File: metadataserver/identity.go [hunk 3]
```diff
@@ -29,6 +31,39 @@ func (ms *MetadataServer) generateTaskIdentity() *titus.TaskIdentity {
...
+	}
+
+	ms.signLock.Lock()
+	oldSigner := ms.signer
+	oldCert, err := x509.ParseCertificate(oldSigner.Certificate.Certificate[0])
+	if err != nil {
+		log.WithError(err).Error("Error parsing old certificate")
+	}
+
+	if newCert != nil && oldCert != nil {
```
Comment (sargun) [inline metadataserver/identity.go]: why not just do a blind 'return' in the `if err != nil` sections? or `x509.ParseCertificate` return errors?

File: cmd/titus-metadata-service/main.go (hunk 1 above)
Comment (rgulewich) [reply cmd/titus-metadata-service/main.go]: Cool, didn't know about `NewTicker()`.

File: metadataserver/identity.go (hunk 3 above)
Comment (rgulewich) [reply metadataserver/identity.go]: I got a little nervous about not doing the assignment when those parses fail.  I guess it's fine, though.

**Comment by rgulewich** [comment]
> Can you wire up SIGHUP as well?

Why, if restarting the systemd unit does the same thing?

**Comment by sargun** [comment]
Because my injection code is jank, and it fails sometimes. Systemd eventually retries restarts and it works again. But, again, jank.

**Comment by rgulewich** [comment]
@sargun - OK, added.

File: cmd/titus-metadata-service/main.go [hunk 4]
```diff
@@ -78,6 +81,29 @@ func readTaskConfigFile(taskID string) (*titus.ContainerInfo, error) {
...
+
+	for {
+		select {
+		case <-t.C:
+		case <-sigs:
+		}
+
+		if newSigner, err := identity.NewDefaultSigner(); err != nil {
+			log.WithError(err).Fatal("Cannot instantiate new default signer")
+		} else {
```
Comment (sargun) [inline cmd/titus-metadata-service/main.go]: Drop the else. The above statement is terminating.

**Pull request merged**